src/generator-context.cs

```diff
@@ -0,0 +1,180 @@
+
+        // How many spans deep is the current context
+        public static int CurrentDepth => Context.Count;
+
+        static void Begin (string context)
+        {
+            Context.Push (context);
+            ShouldPrintContext = true;
+        }
+
+        static void End ()
+        {
+            Context.Pop ();
+            ShouldPrintContext = true;
+        }
+
+        public static SpanOwner BeginSpan (string context)
+        {
+            Begin (context);
+            return new SpanOwner (CurrentDepth);
+        }
+
+        public static TracingOwner TracingEnabled (bool condition)
+        {
+            DetailedTracingEnabled = condition;
+            ShouldPrintContext = true;
+            return new TracingOwner ();
+        }
+
+        public static void EndTracing ()
+        {
+            DetailedTracingEnabled = false;
+            ShouldPrintContext = true;
+        }
+
+        static string Indent (int count) => new string ('\t', count);
+        static string IndentForContext () => new string ('\t', CurrentDepth + 1);
+
+        static void PrintContext ()
+        {
+            if (ShouldPrintContext) {
+                int indent = 0;
+
+                foreach (string span in Context.Reverse()) {
+                    Console.Error.WriteLine (Indent(indent) + span);
+                    indent += 1;
+                }
+                ShouldPrintContext = false;
+            }
+        }
+
+        public static void Trace (object value)
+        {
+            if (DetailedTracingEnabled) {
+                PrintContext ();
+                Console.Error.WriteLine (IndentForContext () + value);
+            }
+        }
+
+        public static void TraceLog (Func<string> logger)
+        {
+            if (DetailedTracingEnabled) {
+                PrintContext ();
+                Console.Error.WriteLine (IndentForContext () + logger());
+            }
+        }
+
+        public static void TraceEnumerable (string title, IEnumerable list)
+        {
+            if (DetailedTracingEnabled) {
+                PrintContext ();
+                Console.Error.WriteLine ($"{IndentForContext ()}{title}:");
+                foreach (var value in list) {
+                    Console.Error.WriteLine ($"{IndentForContext ()}\t{value}");
+                }                
+            }
+        }
+
+        public static void TraceSection ()
+        {
+            if (DetailedTracingEnabled) {
+                PrintContext ();
+                Console.Error.WriteLine ();
+            }
+        }
+
+        public static void TraceStackTrace ()
+        {
+            if (DetailedTracingEnabled) {
+                PrintContext ();
+                StackTrace st = new StackTrace(true);
+                Console.Error.WriteLine ();
+                for (int i = 1; i < st.FrameCount; i++) {
+                    StackFrame? frame = st.GetFrame (i);
+                    if (frame != null) {
```

Comment:
I learned a new patter for this! 
```csharp
if (st.GetFrame(i) is StackFrame frame) 
```
null is not a StackFrame so the code will be the same.